File: public/decode_job_worker.php
<?php

declare(strict_types=1);

require __DIR__ . '/bootstrap.php';

if (PHP_SAPI !== 'cli') {
    fwrite(STDERR, "This script is CLI-only.\n");
    exit(1);
}

$jobId = 0;
foreach ($argv as $i => $arg) {
    if ($arg === '--job-id' && isset($argv[$i + 1])) {
        $jobId = (int) $argv[$i + 1];
        break;
    }
}

if ($jobId <= 0) {
    fwrite(STDERR, "Missing --job-id\n");
    exit(1);
}

try {
    $job = getDecodeJob($jobId);
    if ($job === null) {
        throw new RuntimeException("decode_jobs.id={$jobId} not found");
    }

    $status = (string) ($job['status'] ?? 'queued');
    if (in_array($status, ['succeeded', 'saved', 'running'], true)) {
        exit(0);
    }

    $stmt = pdo()->prepare(
        'UPDATE decode_jobs
         SET status=:status, started_at=NOW(), finished_at=NULL, error_message=NULL, updated_at=CURRENT_TIMESTAMP
         WHERE id=:id'
    );
    $stmt->execute(['id' => $jobId, 'status' => 'running']);

    $sourceImagePath = (string) ($job['source_image_path'] ?? '');
    $promptFilePath = (string) ($job['prompt_file_path'] ?? '');
    if ($sourceImagePath === '' || !is_file($sourceImagePath)) {
        throw new RuntimeException('Source image not found for queued job');
    }
    if ($promptFilePath === '' || !is_file($promptFilePath)) {
        throw new RuntimeException('Prompt snapshot file not found for queued job');
    }

    $requestedModel = trim((string) ($job['requested_model'] ?? ''));
    $result = decodeImageWithPythonPrompt($sourceImagePath, $promptFilePath, $requestedModel !== '' ? $requestedModel : null);
    $decoded = is_array($result['decoded'] ?? null) ? $result['decoded'] : [];
    $usage = is_array($result['usageMetadata'] ?? null) ? $result['usageMetadata'] : [];
    $model = trim((string) ($result['model'] ?? $requestedModel ?: getenv('GEMINI_MODEL') ?: 'gemini-2.5-flash'));
    $totalTokenCount = isset($usage['totalTokenCount']) ? (int) $usage['totalTokenCount'] : null;

    $doneStmt = pdo()->prepare(
        'UPDATE decode_jobs
         SET status=:status,
              decoded_json=:decoded_json,
              usage_metadata_json=:usage_metadata_json,
              decoding_model=:decoding_model,
              total_token_count=:total_token_count,
              finished_at=NOW(),
             updated_at=CURRENT_TIMESTAMP
         WHERE id=:id'
    );
    $doneStmt->execute([
        'id' => $jobId,
        'status' => 'succeeded',
        'decoded_json' => json_encode($decoded, JSON_UNESCAPED_UNICODE),
        'usage_metadata_json' => json_encode($usage, JSON_UNESCAPED_UNICODE),
        'decoding_model' => $model,
        'total_token_count' => $totalTokenCount,
    ]);
    @unlink($promptFilePath);
    exit(0);
} catch (Throwable $e) {
    try {
        $failStmt = pdo()->prepare(
            'UPDATE decode_jobs
             SET status=:status, error_message=:error_message, finished_at=NOW(), updated_at=CURRENT_TIMESTAMP
             WHERE id=:id'
        );
        $failStmt->execute([
            'id' => $jobId,
            'status' => 'failed',
            'error_message' => substr($e->getMessage(), 0, 2000),
        ]);
    } catch (Throwable) {
    }

    fwrite(STDERR, $e->getMessage() . PHP_EOL);
    exit(1);
}

File: python/decode_job_worker.py
#!/usr/bin/env python3
import argparse
import json
import os
import sys
from pathlib import Path

import mysql.connector

import decode_cards


def db_connection():
    return mysql.connector.connect(
        host=os.getenv("MYSQL_HOST", "127.0.0.1"),
        port=int(os.getenv("MYSQL_PORT", "3306")),
        database=os.getenv("MYSQL_DATABASE", "kakracards"),
        user=os.getenv("MYSQL_USER", "root"),
        password=os.getenv("MYSQL_PASSWORD", ""),
    )


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Decode one queued job")
    parser.add_argument("--job-id", type=int, required=True, help="Decode job id")
    return parser.parse_args()


def load_job(cur, job_id: int):
    cur.execute(
        """
        SELECT id, source_image_path, prompt_file_path, requested_model, status
        FROM decode_jobs
        WHERE id=%s
        LIMIT 1
        """,
        (job_id,),
    )
    return cur.fetchone()


def main() -> int:
    decode_cards.load_env()
    args = parse_args()

    conn = db_connection()
    cur = conn.cursor(dictionary=True)

    try:
        job = load_job(cur, args.job_id)
        if not job:
            raise RuntimeError(f"decode_jobs.id={args.job_id} not found")

        if job["status"] in {"succeeded", "saved", "running"}:
            return 0

        cur.execute(
            """
            UPDATE decode_jobs
            SET status='running', started_at=NOW(), finished_at=NULL, error_message=NULL, updated_at=NOW()
            WHERE id=%s
            """,
            (args.job_id,),
        )
        conn.commit()

        prompt_file = str(job.get("prompt_file_path") or "")
        prompt_text = decode_cards.read_prompt(prompt_file)
        requested_model = str(job.get("requested_model") or "").strip() or None
        result = decode_cards.call_gemini(
            Path(str(job["source_image_path"])).resolve(),
            prompt_text,
            requested_model,
        )
        decoded = result["decoded"]
        usage_metadata = result["usageMetadata"]
        model = str(result.get("model") or requested_model or os.getenv("GEMINI_MODEL", "gemini-2.5-flash")).strip()
        total_token_count = int(usage_metadata.get("totalTokenCount") or 0)

        cur.execute(
            """
            UPDATE decode_jobs
            SET status='succeeded',
                decoded_json=%s,
                usage_metadata_json=%s,
                decoding_model=%s,
                total_token_count=%s,
                finished_at=NOW(),
                updated_at=NOW()
            WHERE id=%s
            """,
            (
                json.dumps(decoded, ensure_ascii=False),
                json.dumps(usage_metadata, ensure_ascii=False),
                model,
                total_token_count,
                args.job_id,
            ),
        )
        conn.commit()
        if prompt_file:
            try:
                os.unlink(prompt_file)
            except OSError:
                pass
        return 0

    except Exception as exc:
        cur.execute(
            """
            UPDATE decode_jobs
            SET status='failed', error_message=%s, finished_at=NOW(), updated_at=NOW()
            WHERE id=%s
            """,
            (str(exc)[:2000], args.job_id),
        )
        conn.commit()
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    finally:
        cur.close()
        conn.close()


if __name__ == "__main__":
    raise SystemExit(main())
